### ISSUE TITLE
[FR] model.options["NamedOptions"] needs "RedefineThumbnailOnSave" key

### ISSUE THREAD

**Issue by DanRathbun** [SketchUp/api-issue-tracker#14]
# Feature Request

## SketchUp APIs - Ruby and C

>## **1. Use-case**:

SKP file thumbnail image control plugin needs to check and be sure "**Redefine thumbnail on save**" is **`true`** during model save (via a ModelObserver callback.)

This plugin has been "on the shelf" because of API issues (bugs and missing features) since the mid v7 days! (This is the only missing feature left preventing the plugin from release.)

>## **2. Specific implementation and reasoning**:

The GUI "RedefineThumbnailOnSave" checkbox control is located on the **Model Info** > **File** panel.

### **Ruby**

Although two other controls here on the **Model Info** > **File** panel, (**Name** and **Description**,) are implemented as **`Sketchup::Model`** instance (getter and setter) methods, this particular "option" is best (and likely easiest) to be implemented in the `"NamedOptions"` [**`OptionsProvider`**](http://ruby.sketchup.com/Sketchup/OptionsProvider.html) of the  [**`OptionsManager`**](http://ruby.sketchup.com/Sketchup/OptionsManager.html) interface.
Reasoning is that it's likely not often used enough to warrant the cost of implementing, documenting and maintaining as two separate getter and setter methods. A "property" of the hash-like options collection is more appropriate. The `"NamedOptions"` provider is the most appropriate as this requested key does not "fit" in any of the other providers.

Example of access via Ruby:
```ruby
# Within a ModelObserver subclass object ...
def onPreSaveModel(model)
  # switch off page transitions
  opts = model.options["PageOptions"]
  @pgtrans = opts["ShowTransition"]
  opts["ShowTransition"]= false if @pgtrans
  # activate thumbnail page (user has previously chosen)
  pagename = model.pages.get_attribute("Plugin_Thumbnailer","Thumbnail")
  model.pages.selected_page= pagename
  opts = model.options["NamedOptions"]
  @rtos = opts["RedefineThumbnailOnSave"]
  opts["RedefineThumbnailOnSave"]= true if !@rtos
  # callback returns and SketchUp saves the model ...
end

def onPostSaveModel(model)
  opts = model.options["PageOptions"]
  if opts["ShowTransition"] != @pgtrans
    # Restore previous setting of show transitions
    opts["ShowTransition"]= @pgtrans
  end
  opts = model.options["NamedOptions"]
  if opts["RedefineThumbnailOnSave"] != @rtos
    # Restore previous setting of checkbox
    opts["RedefineThumbnailOnSave"]= @rtos
  end
end
```

### **C API** 
The appropriate C functions already exist, they'd just need to be updated internally to so that the `"NamedOptions"` provider will respond _correctly_ to a `"RedefineThumbnailOnSave"` key:

_See, C API online documentation_: [options_provider.h](http://extensions.sketchup.com/developer_center/sketchup_c_api/sketchup/options__provider_8h_source.html)

~

**Comment by DanRathbun** [comment]
Of course, if SketchUp had better thumbnail scene / saving support built in natively, then I wouldn't even have to have a plugin to workaround the SKP thumbnail pain in the rear-end.

Ie, the **File** panel of the **Model Info** dialog should have a drop list next to the "Redefine Thumbnail" checkbox, in which the user could choose from any currently defined scene pages to be the "Thumbnail" scene. When this control is not empty, then SketchUp should use that scene and it's styling, etc., to create the thumbnail on save regardless of the current scene. This thumbnail save should be transparent to the user (without scene transition delay etc.)

Since this plugin cannot be backward compatible anyway, ... if this thumbnail control (_described above_,) became native, then we wouldn't need to change this setting. It still might be valuable to access it's value if a plugin needed to know if the thumbnail was being updated on save, ... (for example in order to also save an external thumbnail image file for a web gallery or whatever.)

**Comment by thomthom** [comment]
Logged as SU-38591

**Comment by DanRathbun** [comment]
Five years on ... another coder has run into the lack of this feature with the C API:
https://forums.sketchup.com/t/redefine-thumbnail-on-save/186031

**Comment by 3dmod** [comment]
SketchUp crashes a lot during 'Generate Scene Thumbnails'.

 I never use Thumbnails and need a way to set the "Redefine thumbnail on save".

**Comment by thomthom** [comment]
@3dmod - can you log an issue for these crashes?

**Comment by 3dmod** [comment]
ISS-0290 SU Generate Scene Thumbnail Crash

**Comment by thomthom** [comment]
@3dmod for issues in released versions of SketchUp we recommend using this issue tracker as it increase visibility for the API developer community.

**Comment by 3dmod** [comment]
I try to only put API related issues on the Github API issue tracker. 

My Crashing issue is not API related.

Right now my only workaround is to go back to an older version of the SKP.

My request is to move forward the API because it is not the first time I have had an issue with Thumbnail generation and would like a work around.
